libavutil/hwcontext_vaapi.c

```diff
@@ -683,6 +760,26 @@ static int vaapi_get_buffer(AVHWFramesContext *hwfc, AVFrame *frame)
     frame->width   = hwfc->width;
     frame->height  = hwfc->height;
 
+    if (avfc->enable_sub_frame) {
+        if (!ctx->sub_frames_ref)
+            return AVERROR(ENOSYS);
+
+        sub_frame = av_sub_frame_create_side_data(frame);
+        if (!sub_frame)
+            return AVERROR(ENOMEM);
+
+        sub_frames_ctx = (AVHWFramesContext*)ctx->sub_frames_ref->data;
+        sub_frame->buf[0] = av_buffer_pool_get(sub_frames_ctx->pool);
```

Comment:
You can call av_hwframe_get_buffer() to resue vaapi_get_buffer() and the rest code are not needed.